README.md
![Screenshot](screenshot-1.png)

## Overview ##

This is a modular extension for [graphiques-digitale/silverstripe-seo-metadata](https://github.com/Graphiques-Digitale/silverstripe-seo-metadata)

It enables enhanced **_favicon_** and **_pinned icon_** features.

Favicon inspired by: [audreyr/favicon-cheat-sheet][4], [Jonathan T. Neal - Understand the Favicon][5] and [High Quality Visuals for Pinned Sites in Windows 8][6]

It requires:
* [`Graphiques-Digitale/silverstripe-seo-metadata`](https://github.com/Graphiques-Digitale/silverstripe-seo-metadata)

It is intended to be used alongside it's siblings:
* [`Graphiques-Digitale/silverstripe-seo-facebook-domain-insights`](https://github.com/Graphiques-Digitale/silverstripe-seo-facebook-domain-insights)
* [`Graphiques-Digitale/silverstripe-seo-open-graph`](https://github.com/Graphiques-Digitale/silverstripe-seo-open-graph)

These are all optional and fragmented from the alpha version [`SSSEO`](https://github.com/Graphiques-Digitale/SSSEO), which is now redundant.

The whole module collection is based largely on [18 Meta Tags Every Webpage Should Have in 2013][1].

Touch icons inspired by: [Everything you always wanted to know about touch icons][11]

Also, a good overview: [5 tips for SEO with Silverstripe 3][2].

## Installation ##

#### Composer ####

* `composer require graphiques-digitale/silverstripe-seo-icons`
* rebuild using `/dev/build/?flush`

#### From ZIP ####

* Place the extracted folder `silverstripe-seo-icons-{version}` into `silverstripe-seo-icons` in the SilverStripe webroot
* rebuild using `/dev/build/?flush`

## CMS Usage ##

All icons are set globally via `/admin/settings/` under the Metadata tab.

It is recommended you simply have a `favicon.ico` file in the root of your website

## Template Usage ##

Depending on your configuration, the general idea is to replace all header content relating to metadata with `$Metadata()` just below the opening `<head>` tag and `<% base_tag %>` include, e.g.:

```html
<head>
	<% base_tag %>
	$Metadata()
	<!-- further includes ~ viewport, etc. -->
</head>
```

This will output something along the lines of:

```html
<head>
	<base href="http://dev.seo.silverstripe.org/"><!--[if lte IE 6]></base><![endif]-->
	
<!-- SEO -->
<!-- Metadata -->
<meta charset="UTF-8" />
<link rel="canonical" href="http://dev.seo.silverstripe.org/" />
<title>Your Site Name | Home - your tagline here</title>
<meta name="description" content="Welcome to SilverStripe! This is the default home page. You can edit this page by opening the CMS. You can now access the developer documentation, or begin the tutorials." />
<!-- Favicon -->
<link rel="icon" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/logo.png" />
<!--[if IE]><link rel="shortcut icon" href="/favicon.ico" /><![endif]-->
<meta name="msapplication-TileColor" content="FFFFFF" />
<meta name="msapplication-TileImage" content="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/logo.png" />
<!-- Touch Icon -->
<link rel="icon" sizes="192x192" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/_resampled/SetSize192192-logo.png">
<link rel="apple-touch-icon-precomposed" sizes="180x180" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/_resampled/SetSize180180-logo.png">
<link rel="apple-touch-icon-precomposed" sizes="152x152" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/logo.png">
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/_resampled/SetSize144144-logo.png">
<link rel="apple-touch-icon-precomposed" sizes="120x120" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/_resampled/SetSize120120-logo.png">
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/_resampled/SetSize114114-logo.png">
<link rel="apple-touch-icon-precomposed" sizes="76x76" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/_resampled/SetSize7676-logo.png">
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/_resampled/SetSize7272-logo.png">
<link rel="apple-touch-icon-precomposed" href="http://dev.seo.silverstripe.org/assets/SiteConfig/seo-icons/_resampled/SetSize5757-logo.png"><!-- 57×57px -->
<!-- END SEO -->

	<!-- further includes ~ viewport, etc. -->
</head>
```

## Issue Tracker ##

Issues are tracked on GitHub @ [Issue Tracker](https://github.com/Graphiques-Digitale/silverstripe-seo-icons/issues)

## Development and Contribution ##

Please get in touch @ [`hello@graphiquesdigitale.net`](mailto:hello@graphiquesdigitale.net) if you have any extertise in any of these SEO module's areas and would like to help ~ they're a lot to maintain, they should be improved continually as HTML evolves and I'm sure they can generally be improved upon by field experts.

![Screenshot](screenshot-2.png)
![Screenshot](screenshot-3.png)


[1]: https://www.iacquire.com/blog/18-meta-tags-every-webpage-should-have-in-2013
[2]: http://www.silverstripe.org/blog/5-tips-for-seo-with-silverstripe-3-/
[3]: http://moz.com/learn/seo/title-tag
[4]: https://github.com/audreyr/favicon-cheat-sheet
[5]: http://www.jonathantneal.com/blog/understand-the-favicon/
[6]: http://blogs.msdn.com/b/ie/archive/2012/06/08/high-quality-visuals-for-pinned-sites-in-windows-8.aspx
[7]: https://developers.facebook.com/docs/platforminsights/domains
[8]: http://ogp.me
[9]: https://dev.twitter.com/cards/overview
[10]: https://developers.google.com/+/web/snippet/
[11]: https://mathiasbynens.be/notes/touch-icons


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/Graphiques-Digitale/silverstripe-seo-icons/tree/HEAD)

**Merged pull requests:**

- Added standard Scrutinizer config [\#1](https://github.com/Graphiques-Digitale/silverstripe-seo-icons/pull/1) ([helpfulrobot](https://github.com/helpfulrobot))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

code/SEO_Icons_SiteTree_DataExtension.php
<?php

/**
 * @todo Description
 *
 * @package silverstripe-seo
 * @subpackage icons
 * @author Andrew Gerber <atari@graphiquesdigitale.net>
 * @version 1.0.0
 *
 */
class SEO_Icons_SiteTree_DataExtension extends DataExtension
{


    /* Overload Variable
    ------------------------------------------------------------------------------*/

    // none


    /* Overload Methods
    ------------------------------------------------------------------------------*/

    // CMS Fields
// 	public function updateCMSFields(FieldList $fields) {}


    /* Template Methods
    ------------------------------------------------------------------------------*/

    /**
     * @name updateMetadata
     *
     * Updates metadata with icons.
     *
     * @param SiteConfig $config
     * @param SiteTree $owner
     * @param $metadata
     * @return void
     */
    public function updateMetadata(SiteConfig $config, SiteTree $owner, &$metadata)
    {

        //// HTML4 Favicon

        // @todo Perhaps create dynamic image, but just use favicon.ico for now

        //// Create Favicons

        $HTML5Favicon = $config->HTML5Favicon();
        $IOSPinicon = $config->IOSPinicon();
        $AndroidPinicon = $config->AndroidPinicon();
        $WindowsPinicon = $config->WindowsPinicon();

        //// iOS Pinicon

        if ($IOSPinicon->exists()) {

            // header
            $metadata .= $owner->MarkupComment('iOS Pinned Icon');

            //// iOS Pinicon Title
            if ($config->fetchPiniconTitle()) {
                $metadata .= $owner->MarkupMeta('apple-mobile-web-app-title', $config->fetchPiniconTitle());
            }

            //// iOS Pinned Icon

            // For non-Retina (@1× display) iPhone, iPod Touch, and Android 2.1+ devices
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(57, 57)->getAbsoluteURL(), 'image/png'); // 57×57

            // @todo: What is this for ??
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(60, 60)->getAbsoluteURL(), 'image/png', '60x60');

            // For the iPad mini and the first- and second-generation iPad (@1× display) on iOS ≤ 6
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(72, 72)->getAbsoluteURL(), 'image/png', '72x72');

            // For the iPad mini and the first- and second-generation iPad (@1× display) on iOS ≥ 7
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(76, 76)->getAbsoluteURL(), 'image/png', '76x76');

            // For iPhone with @2× display running iOS ≤ 6
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(114, 114)->getAbsoluteURL(), 'image/png', '114x114');

            // For iPhone with @2× display running iOS ≥ 7
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(120, 120)->getAbsoluteURL(), 'image/png', '120x120');

            // For iPad with @2× display running iOS ≤ 6
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(144, 144)->getAbsoluteURL(), 'image/png', '144x144');

            // For iPad with @2× display running iOS ≥ 7
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(152, 152)->getAbsoluteURL(), 'image/png', '152x152');

            // For iPhone 6 Plus with @3× display
            $metadata .= $owner->MarkupLink('apple-touch-icon', $IOSPinicon->SetSize(180, 180)->getAbsoluteURL(), 'image/png', '180x180');

        } else {

            // @todo: hmm??
//			// disabled header
//			$metadata .= $owner->MarkupComment('iOS Pinned Icon DISABLED');

        }

        //// Android Pinicon

        // @todo: Should `rel="icon" sizes="192x192"` be the HTML5 icon or the Android icon ??

//		if ($AndroidPinicon->exists()) {
//
//			// Android Chrome 37+
//			$metadata .= $owner->MarkupLink('icon', $AndroidPinicon->SetSize(192,192)->getAbsoluteURL(), 'image/png', '192x192');
//
//		}

        //// HTML5 Favicon

        if ($HTML5Favicon->exists()) {

            // header
            $metadata .= $owner->MarkupComment('HTML5 Favicon');

//			// Android Chrome 32
            // @todo: Is the Android Chrome 32 196x196 px icon fully redundant ??
//			$metadata .= $owner->MarkupLink('icon', $HTML5Favicon->SetSize(196,196)->getAbsoluteURL(), 'image/png', '196x196');

            // Android Chrome 37+ / HTML5 spec
            $metadata .= $owner->MarkupLink('icon', $HTML5Favicon->SetSize(192, 192)->getAbsoluteURL(), 'image/png', '192x192');

            // Android Chrome 37+ / HTML5 spec
            $metadata .= $owner->MarkupLink('icon', $HTML5Favicon->SetSize(128, 128)->getAbsoluteURL(), 'image/png', '128x128');

            // For Google TV
            $metadata .= $owner->MarkupLink('icon', $HTML5Favicon->SetSize(96, 96)->getAbsoluteURL(), 'image/png', '96x96');

            // For Safari on Mac OS
            $metadata .= $owner->MarkupLink('icon', $HTML5Favicon->SetSize(32, 32)->getAbsoluteURL(), 'image/png', '32x32');

            // The classic favicon, displayed in the tabs
            $metadata .= $owner->MarkupLink('icon', $HTML5Favicon->SetSize(16, 16)->getAbsoluteURL(), 'image/png', '16x16');

        }

        //// Android Pinicon Manifest

        if ($AndroidPinicon->exists()) {

            // header
            $metadata .= $owner->MarkupComment('Android Pinned Icon');

            //
            if ($config->fetchAndroidPiniconThemeColor()) {
                $metadata .= $owner->MarkupMeta('theme-color', $config->fetchAndroidPiniconThemeColor());
            }

            //
            $metadata .= $owner->MarkupLink('manifest', '/manifest.json');

        }

        //// Windows Pinicon Manifest

        if ($WindowsPinicon->exists()) {

            // header
            $metadata .= $owner->MarkupComment('Windows Pinned Icon');

            // application name
            $appName = $config->fetchPiniconTitle();
            if (!$appName) {
                $appName = $config->Title;
            }
            $metadata .= $owner->MarkupMeta('application-name', $appName);

            // tile background color
            if ($config->fetchWindowsPiniconBackgroundColor()) {
                $metadata .= $owner->MarkupMeta('msapplication-TileColor', $config->fetchWindowsPiniconBackgroundColor());
            }

            // small tile
            $metadata .= $owner->MarkupMeta('msapplication-square70x70logo', $WindowsPinicon->Fill(70, 70)->getAbsoluteURL());

            // medium tile
            $metadata .= $owner->MarkupMeta('msapplication-square150x150logo', $WindowsPinicon->Fill(150, 150)->getAbsoluteURL());

            // @todo: Implement wide & tall tiles

            // wide tile
//			$metadata .= $owner->MarkupMeta('msapplication-square310x150logo', $WindowsPinicon->Fill(310,150)->getAbsoluteURL());

            // large tile
//			$metadata .= $owner->MarkupMeta('msapplication-square310x310logo', $WindowsPinicon->Fill(310,310)->getAbsoluteURL());

        }

    }

}


code/SEO_Icons_SiteConfig_DataExtension.php
<?php

/**
 * @todo Description
 *
 * I called them Pinicons & Tabicons in keeping with the absolutely pointless name: Favicons.
 *
 * @package silverstripe-seo
 * @subpackage icons
 * @author Andrew Gerber <atari@graphiquesdigitale.net>
 * @version 1.0.0
 *
 */
class SEO_Icons_SiteConfig_DataExtension extends DataExtension
{


    /* Static Variables
    ------------------------------------------------------------------------------*/

    //
    private static $SEOIconsUpload = 'SEO/Icons/';

    //
    const APPLE_ICON_DEFAULT_BACKGROUND = '000000';


    /* Overload Model
    ------------------------------------------------------------------------------*/

    private static $db = array(
        // Pinned Icon Title
        'PiniconTitle' => 'Varchar(128)',
        // IOS Pinned Icon
//		'IOSPiniconBackgroundColor' => 'Varchar(6)',
        // Android Pinned Icon
        'AndroidPiniconThemeColor' => 'Varchar(6)',
        // MS Tile
        'WindowsPiniconBackgroundColor' => 'Varchar(6)',
        // Safari Pinned Tab
//		'SafariTabiconThemeColor' => 'Varchar(6)', // @todo maybe implement
    );
    private static $has_one = array(
        // Favicon
//		'HTML4Favicon' => 'Image', // @todo maybe implement
        'HTML5Favicon' => 'Image',
        // Apple Pinned Icon
        'IOSPinicon' => 'Image',
        // Android Pinned Icon
        'AndroidPinicon' => 'Image',
        // Windows Pinned Icon
        'WindowsPinicon' => 'Image',
        // Safari Pinned Tab
//		'SafariTabiconDefault' => 'Image', // @todo maybe implement
//		'SafariTabiconWide' => 'Image', // @todo maybe implement
//		'SafariTabiconLarge' => 'Image', // @todo maybe implement
    );


    /* Overload Methods
    ------------------------------------------------------------------------------*/

    // CMS Fields
    public function updateCMSFields(FieldList $fields)
    {

        // owner
        $config = SiteConfig::current_site_config();
        $owner = $this->owner;

        //// Favicons Tab

        $tab = 'Root.Metadata.Favicons';

        //// Favicon

        $fields->addFieldsToTab($tab, array(
            LabelField::create('FaviconDescription', 'Favicons are `favourite icons` used by browsers in a number of ways whenever an icon is necessary e.g. tabs, history, bookmarks, dashboards.<br />@ <a href="https://en.wikipedia.org/wiki/Favicon" target="_blank">Favicon - Wikipedia, the free encyclopedia</a>')
                ->addExtraClass('information')
        ));

        //// HTML4 Favicon

        // check favicon.ico & set status
        $icoStatus = ReadonlyField::create('HTML4FaviconStatus', 'Favicon ICO<pre>type: ico</pre><pre>size: (multiple)<br />16x16 & 32x32 & 64x64 px</pre>', 'favicon.ico error');
        if (Director::fileExists('favicon.ico')) {
            $icoStatus
                ->setValue('favicon.ico found')
                ->addExtraClass('success favicon');
        } else {
            $icoStatus
                ->setValue('favicon.ico not found')
                ->addExtraClass('error');
        }

        // header & fields
        $fields->addFieldsToTab($tab, array(
            HeaderField::create('HTML4FaviconHeader', 'HTML4 <span class="aka">( favicon.ico )</span>'),
            LabelField::create('HTML4FaviconDescription', 'It is recommended you simply have a `favicon.ico` file in the webroot.')
                ->addExtraClass('information'),
            $icoStatus
        ));

        //// HTML5 Favicon

        // header & fields
        $fields->addFieldsToTab($tab, array(
            HeaderField::create('HTML5FaviconHeader', 'HTML5 <span class="aka">( favicon.png )</span>'),
            LabelField::create('HTML5FaviconDescription', '@todo Description')
                ->addExtraClass('information'),
            UploadField::create('HTML5Favicon', 'Favicon PNG<pre>type: png</pre><pre>size: 192x192 px</pre>')
                ->setAllowedExtensions(array('png'))
                ->setFolderName('SiteConfig/seo-icons/')
        ));

        //// Pinned Icons Tab

        $tab = 'Root.Metadata.PinnedIcons';

        //// Pinned Icons Information

        $fields->addFieldsToTab($tab, array(
            LabelField::create('PiniconDescription', 'Pinned icons are OS-specific desktop shortcuts to pages on your website, they allow you to configure additional theming options to make pages appear more `native` / `web-app-y` within the OS.<br />Given they are OS-specific, they (obviously!) have a different format for each one :(')
                ->addExtraClass('information')
        ));

        //// Pinned Icon Title

        // CMS fields
        $fields->addFieldsToTab($tab, array(
            // header
            HeaderField::create('PiniconTitleHeader', 'Pinned Icon Title <span class="aka">( a.k.a. App Name )</span>'),
            // description
            LabelField::create('PiniconTitleDescription', 'When adding a link to the home screen, the user can choose a caption. By default, this is the bookmarked page title, which is usually fine. However, iOS and Windows 8 let you override this default value.')
                ->addExtraClass('information'),
            TextField::create('PiniconTitle', 'Application Title')
                ->setAttribute('placeholder', 'default: page title')
        ));

        //// iOS Pinned Icon

        // CMS fields
        $fields->addFieldsToTab($tab, array(
            // header
            HeaderField::create('IOSPiniconHeader', 'iOS Pinned Icon <span class="aka">( a.k.a. Touch Icons, Web Clips )</span>'),
            // information
            LabelField::create('IOSPiniconDescription', 'iPhone and iPad users can pin your web site on their home screen. The link looks like a native app.<br />@ <a href="https://developer.apple.com/library/ios/documentation/AppleApplications/Reference/SafariWebContent/ConfiguringWebApplications/ConfiguringWebApplications.html" target="_blank">Configuring Web Applications - iOS Developer Library</a>')
                ->addExtraClass('information'),
            // icon
            UploadField::create('IOSPinicon', 'iOS Icon<pre>type: png</pre><pre>size: 192x192 px</pre>')
                ->setAllowedExtensions(array('png'))
                ->setFolderName('SiteConfig/seo-icons/')
                ->setDescription('iOS will fill the transparent regions with black by default, so put your own background in!'),
            // background
//			TextField::create('IOSPiniconBackgroundColor', 'Background Color<pre>type: hex triplet</pre>')
//				->setAttribute('placeholder', self::APPLE_ICON_DEFAULT_BACKGROUND)
//				->setAttribute('size', 6)
//				->setMaxLength(6)
//				->setDescription('iOS will fill the transparent regions with black by default, so put your own background in.')
        ));

        //// Android Pinned Icon

        // CMS fields
        $fields->addFieldsToTab($tab, array(
            // header
            HeaderField::create('AndroidPiniconHeader', 'Android Pinned Icon <span class="aka">( a.k.a. Android Chrome Icons, Launcher Icons )</span>'),
            // information
            LabelField::create('AndroidPiniconDescription', 'Add to Homescreen is a also a feature of Android Chrome. Your visitors can mix their natives apps and web bookmarks.<br />@ <a href="https://developer.chrome.com/multidevice/android/installtohomescreen" target="_blank">Add to Homescreen - Google Chrome</a>')
                ->addExtraClass('information'),
            // icon
            UploadField::create('AndroidPinicon', 'Android Icon<pre>type: png</pre><pre>size: 192x192 px</pre>')
                ->setAllowedExtensions(array('png'))
                ->setFolderName('SiteConfig/seo-icons/'),
            // background
            TextField::create('AndroidPiniconThemeColor', 'Theme Color<pre>type: hex triplet</pre>')
                ->setAttribute('placeholder', 'none')
                ->setAttribute('size', 6)
                ->setMaxLength(6)
                ->setDescription('Starting with Android Lollipop, you can customize the color of the task bar in the switcher.')
        ));

        //// Windows Pinned Icon

        // CMS fields
        $fields->addFieldsToTab($tab, array(
            // header
            HeaderField::create('WindowsShortcutHeader', 'Windows Pinned Icon <span class="aka">( a.k.a. Windows 8 / Metro Tiles )</span>'),
            // information
            LabelField::create('WindowsShortcutDescription', 'Windows 8 users can pin your web site on their desktop. Your site appears as a tile, just like a native Windows 8 app.<br />@ <a href="https://msdn.microsoft.com/en-us/library/dn455106" target="_blank">Creating custom tiles for IE11 websites (Windows)</a>')
                ->addExtraClass('information'),
            // icon
            UploadField::create('WindowsPinicon', 'Windows Icon<pre>type: png</pre><pre>size: 192x192 px</pre>')
                ->setAllowedExtensions(array('png'))
                ->setFolderName('SiteConfig/seo-icons/'),
            // background
            TextField::create('WindowsPiniconBackgroundColor', 'Background ( Tile ) Color<pre>type: hex triplet</pre>')
                ->setAttribute('placeholder', 'none')
                ->setAttribute('size', 6)
                ->setMaxLength(6)
        ));

        // @todo Safari Pinned Tab ~ maybe ??

    }

    public function onAfterWrite()
    {

        // parent
        parent::onAfterWrite();

        // regenerate manifest
        if ($this->generateAndroidManifest()) {
            // @todo SilverStripe success message
        } else {
            // @todo SilverStripe failure message
        }

    }


    /* Custom Methods
    ------------------------------------------------------------------------------*/

    //// fetch functions

    public function fetchPiniconTitle()
    {

        if ($this->owner->PiniconTitle) {
            // return pinicon title
            return $this->owner->PiniconTitle;
        } else {
            // default
            return false;
        }

    }

    public function fetchAndroidPiniconThemeColor()
    {

        if ($this->owner->AndroidPiniconThemeColor) {
            return '#' . $this->owner->AndroidPiniconThemeColor;
        } else {
            return false;
        }

    }

    public function fetchWindowsPiniconBackgroundColor()
    {

        if ($this->owner->WindowsPiniconBackgroundColor) {
            return '#' . $this->owner->WindowsPiniconBackgroundColor;
        } else {
            return false;
        }

    }

    //// generate functions

    public function generateAndroidManifest()
    {

        //// Android Pinicon Manifest

        $pinicon = $this->owner->AndroidPinicon();

        if ($pinicon->exists()) {

            //
            $manifest = new stdClass();

            //
            $manifest->name = $this->owner->PiniconTitle;
//			$manifest->start_url = null; @todo Maybe implement
//			$manifest->display = null; @todo Maybe implement
//			$manifest->orientation = null; @todo Maybe implement
            $manifest->icons = array();

            // 0.75x density icon
            array_push($manifest->icons, array(
                'src' => $pinicon->Fill(36, 36)->getAbsoluteURL(),
                'sizes' => '36x36',
                'type' => 'image/png',
                'density' => 0.75
            ));

            // 1x density icon
            array_push($manifest->icons, array(
                'src' => $pinicon->Fill(48, 48)->getAbsoluteURL(),
                'sizes' => '48x48',
                'type' => 'image/png',
                'density' => 1
            ));

            // 1.5x density icon
            array_push($manifest->icons, array(
                'src' => $pinicon->Fill(72, 72)->getAbsoluteURL(),
                'sizes' => '72x72',
                'type' => 'image/png',
                'density' => 1.5
            ));

            // 2x density icon
            array_push($manifest->icons, array(
                'src' => $pinicon->Fill(96, 96)->getAbsoluteURL(),
                'sizes' => '96x96',
                'type' => 'image/png',
                'density' => 2
            ));

            // 3x density icon
            array_push($manifest->icons, array(
                'src' => $pinicon->Fill(144, 144)->getAbsoluteURL(),
                'sizes' => '144x144',
                'type' => 'image/png',
                'density' => 3
            ));

            // 4x density icon
            array_push($manifest->icons, array(
                'src' => $pinicon->Fill(192, 192)->getAbsoluteURL(),
                'sizes' => '192x192',
                'type' => 'image/png',
                'density' => 4
            ));

            // create file
            $bytes = file_put_contents($_SERVER['DOCUMENT_ROOT'] . '/manifest.json', json_encode($manifest));

            //
            if ($bytes !== false) {
                // success
                return true;
            } else {
                // failure
                return false;
            }

        }

    }

}
